File: src/core/approval-rules.js
// ApprovalRules — YAML-based approval rule engine with regex safety
// See: cc-bridge-v3-final-plan.md Section 4

const fs = require('fs');
const yaml = require('js-yaml');
const { minimatch } = require('minimatch');

function hasNestedQuantifiers(pattern) {
  const groupQuantifier = /\([^)]*[+*{][^)]*\)[+*{]/;
  return groupQuantifier.test(pattern);
}

function hasBackReference(pattern) {
  return /\\[1-9]/.test(pattern);
}

class ApprovalRules {
  constructor(rulesPath) {
    this.rules = this.loadRules(rulesPath);
  }

  loadRules(filePath) {
    const raw = yaml.load(fs.readFileSync(filePath, 'utf8'));
    for (const rule of raw.rules) {
      if (rule.command_pattern) {
        if (rule.command_pattern.length > 100) {
          throw new Error(`Rule regex too long (>100): ${rule.command_pattern}`);
        }
        if (hasNestedQuantifiers(rule.command_pattern)) {
          throw new Error(`Rule regex has nested quantifiers: ${rule.command_pattern}`);
        }
        if (hasBackReference(rule.command_pattern)) {
          throw new Error(`Rule regex has back references: ${rule.command_pattern}`);
        }
        try { new RegExp(rule.command_pattern); }
        catch (e) { throw new Error(`Rule regex compilation failed: ${e.message}`); }
      }
    }
    return raw.rules;
  }

  match(toolName, context = {}) {
    for (const rule of this.rules) {
      if (rule.tool !== toolName && rule.tool !== '*') continue;

      if (rule.command_pattern && context.command) {
        const regex = new RegExp(rule.command_pattern);
        if (!regex.test(context.command)) continue;
      }

      if (rule.path_pattern && context.filePath) {
        if (!minimatch(context.filePath, rule.path_pattern)) continue;
      }

      return {
        action: rule.action,
        sensitive: rule.sensitive || false
      };
    }
    return { action: 'require_approval', sensitive: false };
  }
}

module.exports = { ApprovalRules, hasNestedQuantifiers, hasBackReference };

File: hermes-plugin/core/approval_rules.py
"""ApprovalRules — YAML-based approval rule engine with regex safety.

Port of openclaw-plugin/src/core/approval-rules.js.
PyYAML for loading rules; fnmatch for path_pattern matching;
re for command_pattern matching.

See: cc-bridge-v3-final-plan.md Section 4, shared/approval-api.md
"""

import fnmatch
import re
import yaml
from typing import Any


def has_nested_quantifiers(pattern: str) -> bool:
    """Detect nested quantifiers like (a+)+ or (a*)* that cause ReDoS."""
    group_quantifier = re.compile(r"\([^)]*[+*{][^)]*\)[+*{]")
    return bool(group_quantifier.search(pattern))


def has_back_reference(pattern: str) -> bool:
    """Detect back references like \\1, \\2 in regex patterns."""
    return bool(re.search(r"\\[1-9]", pattern))


class ApprovalRules:
    """Load and evaluate approval rules from a YAML file.

    Rules are evaluated in order. First matching rule wins.
    Default action is require_approval (fail-closed).
    """

    def __init__(self, rules_path: str) -> None:
        self.rules: list[dict[str, Any]] = self._load_rules(rules_path)

    def _load_rules(self, file_path: str) -> list[dict[str, Any]]:
        """Load rules from YAML, compile regex, and validate safety constraints."""
        with open(file_path, "r", encoding="utf-8") as fh:
            raw = yaml.safe_load(fh)

        rules = raw.get("rules", [])
        for rule in rules:
            cmd_pattern = rule.get("command_pattern")
            if cmd_pattern:
                if len(cmd_pattern) > 100:
                    raise ValueError(
                        f"Rule regex too long (>100): {cmd_pattern}"
                    )
                if has_nested_quantifiers(cmd_pattern):
                    raise ValueError(
                        f"Rule regex has nested quantifiers: {cmd_pattern}"
                    )
                if has_back_reference(cmd_pattern):
                    raise ValueError(
                        f"Rule regex has back references: {cmd_pattern}"
                    )
                try:
                    rule["_compiled_regex"] = re.compile(cmd_pattern)
                except re.error as exc:
                    raise ValueError(
                        f"Rule regex compilation failed: {exc}"
                    ) from exc

        return rules

    def match(
        self, tool_name: str, context: dict[str, Any] | None = None
    ) -> dict[str, Any]:
        """Match a tool name and context against the rules list.

        Returns {"action": <action>, "sensitive": <bool>}.
        Default if no rule matches: require_approval, sensitive=False.
        """
        if context is None:
            context = {}

        for rule in self.rules:
            # Tool name must match exactly or via * wildcard
            if rule.get("tool") != tool_name and rule.get("tool") != "*":
                continue

            # command_pattern: regex match against context.command
            cmd_pattern = rule.get("command_pattern")
            if cmd_pattern and context.get("command"):
                compiled = rule.get("_compiled_regex")
                if compiled and not compiled.search(context["command"]):
                    continue

            # path_pattern: fnmatch against context.filePath (glob matching)
            path_pattern = rule.get("path_pattern")
            if path_pattern and context.get("filePath"):
                if not fnmatch.fnmatch(context["filePath"], path_pattern):
                    continue

            return {
                "action": rule.get("action", "require_approval"),
                "sensitive": rule.get("sensitive", False),
            }

        # Default: fail-closed
        return {"action": "require_approval", "sensitive": False}